docs/src/internals/ir.md

```diff
@@ -0,0 +1,155 @@
+# Yk Intermediate IR
+
+**This is a WIP document. We are currently in the process of switching from
+using LLVM IR to our own IR. This document describes the new IR**
+
+## On-disk serialisation format.
+
+(syntax loosely based on
+https://www.llvm.org/docs/StackMaps.html#stack-map-format)
+
+```
+header {
+    magic: usize    // hard-coded `0xedd5f00d`
+    version: usize  // format version (currently 0)
+}
+toc {
+    funcs {
+        num_funcs: usize
+        func_offs[num_funcs] {      // indices of this array server as function IDs.
+            func_data_off: usize    // byte offset into `func_data`.
+        }
+        func_data_size: usize       // in bytes.
+                                    // allows deserialiser to quickly skip over
+                                    // `func_data`.
+    }
+    globals {
+        num_globals: usize
+        global_offs[num_globals] {  // indices of this array serve as global IDs.
+            global_data_off: usize  // byte offset into `global_data`.
+        }
+        global_data_size: usize     // in bytes.
+                                    // allows deserialiser to quickly skip over
+                                    // `global_data`.
+    }
+    types {
+        num_types: usize
+        type_offs[num_types] {      // indices of this array serve as type IDs.
+            type_data_off: usize    // byte offset into `type_data`
+        }
+        type_data_size: usize       // in bytes.
+                                    // allows deserialiser to quickly skip over
+                                    // `type_data`.
+    }
+    constants {
+        num_constants: usize
+        type_offs[num_types] {
+            constant_data_off: usize    // byte offset into `constant_data`
+        }
+    }
+}
+func_data[num_funcs] {
+    name: null_term_str
+    func_sig {
+        num_args: usize
+        arg_tys[num_args] {
+            ty_index: usize
+        }
+    }
+    func_index {
+        num_blocks: usize           // value of 0 means this is a declaration.
+        block_offs[num_blocks] {
+            block_data_off: usize   // byte offset into `block_data`.
+        }
+    }
+    block_data[num_blocks] {
+        num_instrs: u32
```

Comment:
Did you mean `usize`? If not, why do we use a `u32` here when we used a `usize` everywhere else.

Reply:
Yes, it was supposed to be a usize.

I toyed with the idea of using u32s so that we could identify a basicblock using only one pointer-sized field (func-idx: u32 + bbidx: u32), but deemed it a) limiting, and b) too clever for a first stab.